lib/includes.js

```diff
@@ -0,0 +1,189 @@
+        return fetchLinked({}, resources, inclusionTree)
+            .then(function (linked) {
+                body.included = linked;
+                return body;
+            });
+
+    }
+
+    function fetchLinked(fetchedIds, resources, inclusionBranch) {
+        return Promise
+            .all(_.map(_.keys(inclusionBranch ? inclusionBranch.refs : []), function (inclusionRefKey) {
+                return fetchResources(fetchedIds, resources, inclusionBranch, inclusionRefKey)
+                    .then(function (result) {
+                        if (result && result.data) {
+                            return fetchLinked(fetchedIds, result.data, inclusionBranch.refs[inclusionRefKey]).then(function (subResult) {
+                                return _.union(result.data, subResult);
+                            });
+                        } else {
+                            return [];
+                        }
+                    });
+            }))
+            .then(function (linkedResources) {
+                return _.reduce(linkedResources, function (acc, linkedResource) {
+                    return _.union(acc, linkedResource);
+                }, []);
+            });
+    }
+
+    function fetchResources(fetchedIds, resources, inclusionBranch, inclusionRefKey) {
+        var linkedIds = getLinkedIds(resources, inclusionRefKey);
+        if (linkedIds && linkedIds.length > 0) {
+
+            var inclusionDescriptor = inclusionBranch.refs[inclusionRefKey];
+            var type = inclusionDescriptor.def.type;
+
+            fetchedIds[type] = fetchedIds[type] || [];
+            var remainingIds = _.without(linkedIds, fetchedIds[type]);
+            fetchedIds[type] = fetchedIds[type].concat(remainingIds);
+
+            if (!inclusionDescriptor.def.baseUri) {
+                return adapter.find(type, {query: {filter: {id: remainingIds.join(',')}}});
+            } else {
+                // the related resource is defined on another domain
+                // fetch with an http call with inclusion of the deep linked resources for this resource
+                var url = inclusionDescriptor.def.baseUri + '/' + type + '/' + remainingIds.join(',');
+                if (!_.isEmpty(inclusionDescriptor.def.remoteIncludes)) {
+                    url += '?include=' + inclusionDescriptor.def.remoteIncludes;
+                }
+                return $http(url, {json: true})
+                    .spread(function (response, body) {
+                        var result = body.included || [];
+                        result.push(body.data);
+                        return {data: result};
+                    });
+            }
+
+        } else {
+            return Promise.resolve();
+        }
+    }
+
+    function getLinkedIds(resources, path) {
+        var ids = _.reduce(resources, function (acc, resource) {
+            if (resource.relationships && resource.relationships[path]) {
```

Comment:
I can't quite see it, does this code conform to JSON API 1.0 relationships? They can never be arrays, their ```data``` property can be.